File: arrays/countNumberOfPairsWithAbsoluteDifferenceK.py
# https://leetcode.com/problems/count-number-of-pairs-with-absolute-difference-k/

from collections import Counter

class Solution(object):
    
    # Brute Force Approach
    
    # TC : O(N
    # SC : O(N)
    def countKDifference(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        count =0
        
        for i in range(len(nums)):
            for j in range(i, len(nums)):
                if abs(nums[i]-nums[j])==k: 
                    count +=1
        return count
                
                
                
    
    def countKDifference(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        dictionary =Counter()
        count=0
        for i in nums:
            dictionary[i]+=1
            count += dictionary[i-k] + dictionary[i+k]
            
        return count 

File: arrays/countNumberOfPairsWithAbsoluteDifferenceK.js
// Runtime: 95 ms, faster than 73.30% of JavaScript online submissions for Count Number of Pairs With Absolute Difference K.
// Memory Usage: 49.7 MB, less than 5.21% of JavaScript online submissions for Count Number of Pairs With Absolute Difference K.

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const countKDifference = (nums, k) => {
    const acc = [];
    for (let i = 0; i < nums.length; i++) {
        helper(k, nums, nums[i], i, acc)
    }
    return acc.length / 2;
};
const helper = (target, arr, val, index, acc) => {
    for (const value of arr) {
        if (Math.abs(val - value) === target) {
            acc.push([val, value])
        }
    }
}
module.exports = countKDifference;